### PR TITLE
Implement named capture groups

### PR DESCRIPTION
**Pull request by raskad**
This PR contains a working implementation of named capture groups and named backreferences.
To validate the functionality, I have added the most relevant named capture group tests from the 262 test suite.
Further I have implemented the named capture groups in fork of [boa](https://github.com/boa-dev/boa) (I can push that code if needed) and have validated that 262 tests also pass there.

The implementation and especially the addition to the public api are done with minimal change in mind, to get this to work. Personally I would prefer if `Match.group` and `Match.groups` would be changed to additionally return a `Option<groupname>` as this seems to be the natural implementation when implementing `RegExp.exec` according to the spec.

In addition to the named capture groups I have implemented a parsing of unicode escape characters e.g. `\u{0041}` and `\u0041`. This seemed necessary to me, as the 262 tests for named capture groups contain a lot of tests with theses escaped characters. The implementation of this parsing is very much not optimal, as I have just implemented a naive parsing of the whole input before any real parsing. I think in theory this parsing could be done on demand in `consume_atom_escape`, but I have not found an easy way to further parse the parsed unicode char after parsing the escape sequence at that point.

@ridiculousfish I would very much like to work on getting these features merged. Please let me know if you think we can work of of this code or what other approaches could look like.

### CODE REVIEW

**Comment by ridiculousfish** [comment]
This looks amazing! Unfortunately I didn't get to reviewing it today but I will ASAP

**Review by ridiculousfish — COMMENTED** [review]
This looks excellent - thank you! It's awesome how the executors don't need to change.

My biggest concern before merging is the `parse_capture_groups` function, which seems to miss some edge cases, so I think it's a correctness issue. I added a suggestion for changing it to a separate pass over the IR.

Some other issues that don't need block merge (you can fix them now or one of us can do it later, at your option):

1. regress wants to be low dependency so I would prefer to not take a dependency on `boa_unicode`; rather have `gen-folds.rs` emit the Unicode property tables.

2. As you acknowledged the new unicode escaping is sub-optimal because it copies the pattern. One way to do it is to wrap the Parser's `input` (currently `Peekable<Chars>`) with an iterator adapter that can parse Unicode escapes. On error it would return None and set an internal error flag.

As I said this looks to be shaping up to an awesome contribution, thank you!

File: tests/tests.rs
Comment (raskad) [reply tests/tests.rs:1007]: Thanks :)

File: src/parse.rs
Comment (raskad) [reply src/parse.rs:846]: > Is the idea pre-populate `named_group_indices` for backreferences?

Yes that is the idea. The challange with doing a pass on the IR is, that without the `unicode` flag, a `\k` can be parsed as the literal `k` character if there are no named capture groups in the entire regex. I actually did not implement the unicode flag handling correctly before. I just pushed aa3d2ac and fixed that error while adding the AnnexB tests for the named capture groups. If you take a look at the new `\k` handling in `consume_atom_escape` ([line 746+](https://github.com/ridiculousfish/regress/commit/aa3d2ac9681ffdd17c1a1d2ceb3becfc7d5010a8#diff-5583e398a8deec11b274d9965eff8b5ade5226c7a020ca214a27d7d07dcc8a29R746)) you may see why I went with this approach.

Concerning the correctness and the approach in general; while implementing I took a look at the v8 implementation. They choose this approach too. While I'm not fluent in c++, I think my parsing function should be identical to the v8 one. See the v8 version here: https://github.com/v8/v8/blob/master/src/regexp/regexp-parser.cc#L728

Because of the wired AnnexB edge cases, I'm not 100% sure if the IR pass approach would be that easy to implement. But I'm also only getting stared on this codebase, so I would like to defer to your judgement on this ;)

Comment (ridiculousfish) [reply src/parse.rs:846]: Nicely found. If v8 implements it this way it's probably fine. I really do think there's a bug with \ handling in character classes: [this specific line](https://github.com/v8/v8/blob/4e1a6c418ad333ad8d2ddccad443bd86ce932c0f/src/regexp/regexp-parser.cc#L751) is the logic that's missing in this diff and I think `[\](.)]` should trigger it. Once that's fixed it should be OK.

Comment (raskad) [reply src/parse.rs:846]: I will do some testing with that example and fix it.

With b444806 I removed the `boa_unicode` dependency and refactored the `gen-folds` crate into a more general `gen-unicode` crate. You might want to take a look at that commit specifically to see the difference.

I'm also working on implementing the unicode escape parsing more closely to the spec. I will ping you once all that work is done.

Comment (raskad) [reply src/parse.rs:962]: Indeed you where right, I did miss the escape handling. I fixed it and wrote a test.

Unicode escape parsing is now done in the main parsing. I think that should conclude all the issues you found.

File: src/classicalbacktrack.rs
Comment (ridiculousfish) [inline src/classicalbacktrack.rs:698]: It is an important invariant that this is not necessary: if you enter a capture group its end must already be None. However this invariant was violated due to a bug!

I fixed the bug in 04a607e7b558919f2da3304a022536afb6ed1a0a so once again this ought not to be necessary. Please rebase on top of master, then replace this `cg.end = None` assignment with something like:

    debug_assert!(cg.end.is_none(), "Should not have already exited capture group we are entering")
    
likewise for `cg.start` below. Anyways great catch.

File: tests/syntax_error_tests.rs
Comment (ridiculousfish) [inline tests/syntax_error_tests.rs:52]: Let's restore this; I don't think we should support the non-Unicode paths where escapes are conditionally backreferences or decimal.

**Review by ridiculousfish — APPROVED** [review]
THIS LOOKS AWESOME! I will extend you commit privileges and you may merge at your leisure.

Two things:

1. Let's not introduce the `u` flag, it's just always assumed to be true.
2. Please rebase on top of master; that will allow the test to pass without adding the `cg.end = None` line (see review comment).

File: src/api.rs
Comment (raskad) [reply src/api.rs]: I removed the 'u' flag and adjusted my new tests accordingly.

Just FYI and for some context; I'm always testing on my local fork of boa to see how my changes affect the ecma262 test suite. With the 'u' flag I get +100 passing tests on this branch, without it I "only" get +54.

Of course you are right. If anything ,this branch is not the right place to introduce something like that. I think at some point in the future I will make an effort to maybe introduce the 'u' flag to make regress and boa more spec compliant. Maybe it could even be a reverse flag ('non-unicode') to introduce 'legacy' behavior. But that is for another time ;)

File: tests/syntax_error_tests.rs
Comment (raskad) [reply tests/syntax_error_tests.rs:52]: I restored the test. The error message is different but the result is the same.

**Comment by raskad** [comment]
> I will extend you commit privileges and you may merge at your leisure.

@ridiculousfish Thank you so much for your trust. I feel honored :)

I have implemented your changes and removed the "Missing Syntax: Named capture groups" mentions from the README and the docs.

**Pull request merged**
